### PR TITLE
8287432: C2: assert(tn->in(0) != __null) failed: must have live top node

### PR DESCRIPTION
**Pull request by chhagedorn**
When intrisifying `java.lang.Thread::currentThread()`, we are creating an `AddP` node that has the `top` node as base to indicate that we do not have an oop (using `NULL` instead leads to crashes as it does not seem to be expected to have a `NULL` base):
https://github.com/openjdk/jdk/blob/6ff2d89ea11934bb13c8a419e7bad4fd40f76759/src/hotspot/share/opto/library_call.cpp#L904

This node is used on a chain of data nodes into two `MemBarAcquire` nodes as precedence edge in the test case:
Screenshot from 2022-06-07 11-12-38

Later, in `final_graph_reshaping_impl()`, we are removing the precedence edge of both `MemBarAcquire` nodes and clean up all now dead nodes as a result of the removal:
https://github.com/openjdk/jdk/blob/6ff2d89ea11934bb13c8a419e7bad4fd40f76759/src/hotspot/share/opto/compile.cpp#L3655-L3679

We iteratively call `disconnect_inputs()` for all nodes that have no output anymore (i.e. dead nodes). This code, however, also treats the `top` node as dead since `outcnt()` of `top` is always zero:
https://github.com/openjdk/jdk/blob/6ff2d89ea11934bb13c8a419e7bad4fd40f76759/src/hotspot/share/opto/node.hpp#L495-L500

And we end up disconnecting `top` which results in the assertion failure.

The code misses a check for `top()`. I suggest to add this check before processing a node for which `outcnt()` is zero. This is a pattern which can also be found in other places in the code. I've checked all other usages of `oucnt() == 0` and could not find a case where this additional `top()` check is missing. Maybe we should refactor these two checks into a single method at some point to not need to worry about `top` anymore in the future when checking if a node is dead based on the outputs.

Thanks,
Christian

---------
### Progress
- [x] Change must be properly reviewed (1 review required, with at least 1 [Reviewer](https://openjdk.java.net/bylaws#reviewer))
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue

### Issue
 * [JDK-8287432](https://bugs.openjdk.org/browse/JDK-8287432): C2: assert(tn->in(0) != __null) failed: must have live top node

### Reviewers
 * [Vladimir Kozlov](https://openjdk.java.net/census#kvn) (@vnkozlov - **Reviewer**) ⚠️ Review applies to [3f351d40](https://git.openjdk.java.net/jdk/pull/9060/files/3f351d4014fc169fa28dede36642ac227f8c7eb0)
 * [Tobias Hartmann](https://openjdk.java.net/census#thartmann) (@TobiHartmann - **Reviewer**) ⚠️ Review applies to [3f351d40](https://git.openjdk.java.net/jdk/pull/9060/files/3f351d4014fc169fa28dede36642ac227f8c7eb0)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.java.net/jdk pull/9060/head:pull/9060` \
`$ git checkout pull/9060`

Update a local copy of the PR: \
`$ git checkout pull/9060` \
`$ git pull https://git.openjdk.java.net/jdk pull/9060/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 9060`

View PR using the GUI difftool: \
`$ git pr show -t 9060`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.java.net/jdk/pull/9060.diff">https://git.openjdk.java.net/jdk/pull/9060.diff</a>

</details>

### CODE REVIEW

**Review by vnkozlov — APPROVED** [review]
Your analysis and fix is correct.

File: test/hotspot/jtreg/compiler/c2/TestRemoveMemBarPrecEdge.java [hunk 1]
```diff
@@ -0,0 +1,55 @@
...
+
+    public static void main(String[] args) {
+        for (int i = 0; i < 10000; i++) {
+            test();
+            flag = !flag;
+        }
+    }
+
+    public static void test() {
+        // currentThread() is intrisified and C2 emits a special AddP node with a base that is top.
```
Comment (TobiHartmann) [inline test/hotspot/jtreg/compiler/c2/TestRemoveMemBarPrecEdge.java:45]:
```suggestion
        // currentThread() is intrinsified and C2 emits a special AddP node with a base that is top.
```

**Review by TobiHartmann — APPROVED** [review]
Nice analysis and test. Looks good.

**Comment by devinrsmith** [comment]
I'm not qualified to review this, but I can confirm the test is triggering the `PhaseAggressiveCoalesce::coalesce` SIGSEGV against the 11/17 LTS versions I'm running. Thanks for your investigation and fix!

**Comment by chhagedorn** [comment]
Thanks @devinrsmith for confirming this!

/integrate

**Comment by devinrsmith** [comment]
@chhagedorn, thanks again! I'm hoping this can be backported to 11/17 - let me know if there is anything I can do to make that happen.

**Pull request closed without merging**